### ISSUE TITLE
bug in alternative parameterization recognition for dmnorm with cholesky as input

### ISSUE THREAD

**Issue by paciorek** [nimble-dev/nimble#841]
```
code <- nimbleCode({
  a[1:2] ~ dmnorm(mu[1:2], cholesky = pr[1:2,1:2])
})
m <- nimbleModel(code, constants = list(mu = c(0,0)), inits = list(pr=matrix(c(1,0,.7,.3),2)))
defining model...
Error in distRule$alts[[count]] : subscript out of bounds
```

**Comment by paciorek** [comment]
Ok, this is simply because I forgot to include `prec_param`. So actually this is a case of needing to trap the error. 

For dmnorm, if one provides cov or prec args, then prec_param is unnecessary...

A couple things:

1) `dmnorm` does not have an altParam for cholesky input, so it can't match it. Only 2 alts despite 3 elements to 'Rdist'.

2) This seems not robust:

```
            while(is.null(matchedAlt) && count < distRule$numAlts) {
              count <- count + 1
              if(identical(sort(unique(distRule$alts[[count]])), sort(unique(names(params)))))
                matchedAlt <- count
            }
```

because `count` is incrementing based on `distRule$numAlts`, but is indexing `distRule$alts`, which is often of length not equal to `distRule$numAlts` (e.g., 'dbin', 'dmnorm').

Still digging...

**Comment by danielturek** [comment]
@paciorek So I spent some time with this.  Thanks for finding in the first place, and the initial diagnosis.  I think the story is a little different, though, but I certainly could be wrong.

Basically, I agree that the calculation of `numAlts` was wrong.  I've fixed that in #843, so that instead that's calculated as `numAlts <- length(alts)`, as it should be.

Your observation about the logical / numeric vector is right, in that the `boolNoDefault` vector was a little mis-named... is it a logical vector?  is it a numeric vector?  So, I also took your suggested change there, verbatim, which clears up that confusion.  However, the difference from what you said is that I think the old code, sometimes setting `boolNoDefault <- seq_along(...)` was sound, in that it (itself) wasn't causing any unwanted downstream behaviour.  e.g., when **no** defaults are provided in an alternate parameterization, (e.g., for `dmnorm(mean, cholesky, prec_param`), then the `names(...)` is `NULL`, and the `boolNoDefault` gets set to `1:3`, which when interpreted as a logical vector, **correctly** indicates that *all arguments of that parameterization have no default values*, as the case is.

And, per your comments about the `if(...)` statement on (former) line 143 "failing", that's not a problem, at all. That correctly indicates that there's no way to reparameterize to this possible parameterization, since all those arguments (in that parameterization) are required -- and hence we can't "reparameterize" to it.  So, in that particular case  (`dmnorm(mean, cholesky, prec_param`), the `numAlts` should end up being 2.

If anything, and I'm just realizing this now as I'm typing it, perhaps that `if(...)` statement should also have an `else` clause, essentially checking that, in that specific situation, the argument names (none of which have default values available) are, in fact, identically the same as the `reqdArgs`, the required set of arguments in our (under-the-hood) canonical parameterization.  Again, I have not fully thought this through, but what I'm suggesting might look something like:

```
if(sum(!boolNoDefault)) {
   ...
} else {
   if(!identical(sort(unlist(paramsText[[i]])), sort(reqdArgs)))
      stop(paste0('reparametization number ', i, ' for ', BUGSdistName, ' with no default argument values must exactly match arguments of canonical parameterization'))
}
```

All that said; I'm not sure if this would actually catch anything, or is necessary; and it's only an error trap / catch anyway, and not mission critical code.

Stepping back.  I think I fixed the problem with `numAlts`.  I suggested an error trap to add.  

And #843 also includes an update where if a user provides the alternate distribution arguments in **different orders**, we'll still handle it correctly.  That's the addition of `lapply(..., sort)` in the line:
```
if(length(unique(lapply(reqdArgsList, sort))) > 1)
```

Ok, maybe a lot to process, but there you go.
